### PR TITLE
Add service worker caching to the example project

### PR DESCRIPTION
**Pull request by jamsea**

### CODE REVIEW

File: example/config/serviceWorkers.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+swPrecache.write(serviceWorkerPath, {
+    cacheId: wildcatConfig.generalSettings.name,
+    staticFileGlobs: [
+        "bundles/*.js",
+        "public/components/**/*",
+        "public/routes/**/*",
+        "public/*.js",
+        "public/assets/styles/*.js",
+        "public/assets/images/*.jpg",
+        "public/application.config.js",
```
Comment (doctyper) [inline example/config/serviceWorkers.js]: I think all of these `public` globs can be reduced to one: `public/**/*`

File: example/config/serviceWorkers.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+const wildcatConfig = require("../wildcat.config.js");
+
+const swDirectory = `${__dirname}/../static`;
+const staticServer = wildcatConfig.serverSettings.staticServer;
+const usePort = staticServer.hostname.indexOf("localhost") !== -1;
+
+let staticHost = `https://${staticServer.hostname}`;
+if (usePort) {
+    staticHost = `${staticHost}:${staticServer.port}`;
+    console.log("USE PORT: ", staticHost);
```
Comment (doctyper) [inline example/config/serviceWorkers.js]: Remove log.

File: example/package.json [hunk 3]
```diff
@@ -105,7 +108,8 @@
     }
   },
   "scripts": {
-    "build": "npm run compile -- --ignore '**/{shell,e2e,test}/**' --minify && npm run bundle",
+    "build": "npm run build-precache && npm run compile -- --ignore '**/{shell,e2e,test}/**' --minify && npm run bundle",
```
Comment (doctyper) [inline example/package.json]: `bundle` needs to run before `build-precache`

File: package.json [hunk 4]
```diff
@@ -85,5 +86,8 @@
     "commitizen": {
       "path": "./node_modules/cz-conventional-changelog"
     }
+  },
+  "dependencies": {
+    "psi": "^2.0.3"
```
Comment (doctyper) [inline package.json]: `psi`?

File: packages/react-wildcat/src/utils/templates/wildcat.config.js [hunk 5]
```diff
@@ -71,6 +71,8 @@ const wildcatConfig = {
         hotReload: !__PROD__,
         hotReloader: "react-wildcat-hot-reloader",
 
+        serviceWorker: false,
```
Comment (doctyper) [inline packages/react-wildcat/src/utils/templates/wildcat.config.js]: Should default to `serviceWorker: __PROD__` here

File: packages/react-wildcat-handoff/src/utils/defaultTemplate.js [hunk 6]
```diff
@@ -41,7 +42,19 @@ module.exports = function defaultTemplate(cfg) {
     </head>
     <body>
         <div id="${reactRootElementID}">${html}</div>
-
+        ${__PROD__ && serviceWorker ? `
```
Comment (doctyper) [inline packages/react-wildcat-handoff/src/utils/defaultTemplate.js]: Should remove `__PROD__` here

File: example/wildcat.config.js [hunk 7]
```diff
@@ -74,6 +74,8 @@ const wildcatConfig = {
 
         hotReload: !__PROD__,
 
+        serviceWorker: true,
```
Comment (doctyper) [inline example/wildcat.config.js]: Use `serviceWorker: __PROD__`

Comment (doctyper) [inline example/wildcat.config.js]: Or just remove this altogether.

File: packages/react-wildcat-handoff/src/utils/defaultTemplate.js (hunk 6 above)
Comment (jamsea) [inline packages/react-wildcat-handoff/src/utils/defaultTemplate.js]: If `__PROD__` is false we want to run the `// Remove any registered service workers (dev mode only)` code, even if serviceWorker is `true` don't we?

File: packages/react-wildcat-handoff/src/utils/defaultTemplate.js [hunk 8]
```diff
@@ -35,13 +36,28 @@ module.exports = function defaultTemplate(cfg) {
...
         <div id="${reactRootElementID}">${html}</div>
-
+        ${serviceWorker ? `
+        <script src="/register-sw.js"></script>
+        ` : `
+        <script>
+            // Remove any registered service workers (dev mode only)
+            navigator.serviceWorker.getRegistrations().then(function(registrations) {
+                for(let registration of registrations) {
+                    console.log("reg: ", registration);
```
Comment (doctyper) [inline packages/react-wildcat-handoff/src/utils/defaultTemplate.js]: Remove log

File: packages/react-wildcat-handoff/src/utils/defaultTemplate.js [hunk 9]
```diff
@@ -35,13 +36,28 @@ module.exports = function defaultTemplate(cfg) {
...
     <body>
         <div id="${reactRootElementID}">${html}</div>
-
+        ${serviceWorker ? `
+        <script src="/register-sw.js"></script>
+        ` : `
+        <script>
+            // Remove any registered service workers (dev mode only)
+            navigator.serviceWorker.getRegistrations().then(function(registrations) {
+                for(let registration of registrations) {
```
Comment (doctyper) [inline packages/react-wildcat-handoff/src/utils/defaultTemplate.js]: Space after for. Use `var` not `let`, we need to dip down to ES5 for certain browsers.

File: packages/react-wildcat-handoff/src/utils/defaultTemplate.js [hunk 10]
```diff
@@ -35,13 +36,28 @@ module.exports = function defaultTemplate(cfg) {
...
+                    console.log("reg: ", registration);
+                    registration.unregister();
+                }
+                if(registrations.length > 0) {
```
Comment (doctyper) [inline packages/react-wildcat-handoff/src/utils/defaultTemplate.js]: Space after `if`

File: packages/react-wildcat-handoff/src/utils/defaultTemplate.js [hunk 11]
```diff
@@ -35,13 +36,27 @@ module.exports = function defaultTemplate(cfg) {
...
     </head>
     <body>
         <div id="${reactRootElementID}">${html}</div>
-
+        ${serviceWorker ? `
+        <script src="/register-sw.js"></script>
+        ` : `
+        <script>
+            // Remove any registered service workers (dev mode only)
+            navigator.serviceWorker.getRegistrations().then(function(registrations) {
```
Comment (doctyper) [inline packages/react-wildcat-handoff/src/utils/defaultTemplate.js]: Can you add a sanity-check for `navigator.serviceWorker`? This code may run in a browser that does not support that feature.

**Pull request merged**